library/std/src/net/endian.rs

```diff
@@ -0,0 +1,38 @@
+//! Define internal endianness primitives for networking.
+
+/// A Big Endian `[u8; 4]`.
+///
+/// Mostly used as inner type of `Ipv4Addr`.
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq)]
+pub(super) struct BeU8x4([u8; 4]);
+
+impl BeU8x4 {
+    /// The argument `n` is interpreted in big endian order.
+    ///
+    /// Caller has to ensure this invariant is satisfied.
+    pub const fn from_be(n: u32) -> Self {
+        // Already in big endian order. See `to_be` for more explanation.
+        Self(n.to_ne_bytes())
+    }
+
+    /// Interprets `bytes` in big endian order.
+    pub const fn from_be_bytes(bytes: [u8; 4]) -> Self {
+        Self(bytes)
+    }
+
+    /// Returns a big endian `u32`.
+    pub const fn to_be(&self) -> u32 {
+        // This is tricky but correct.
+        // For example, given `self.0 = [0xca, 0xfe, 0xba, 0xbe]`:
+        // In little endian target, we have big endian value `0xbebafeca`.
+        // In big endian target, we have value `0xcafebabe` as big endian just like its target endian.
```

Comment:
I still find this confusing. Here's the argument I made for myself: `from_ne_bytes` is really just a transmute between `u32` and `[u8; 4]` without doing any conversion, and that's what we want here. We want nothing to do with endianess.

That's why I think code clarity here is improved by doing a transmute locally instead of calling a method that happens to be a transmute.